bvm/ballerina-core/src/main/java/org/ballerinalang/helpers/DeepEqualHelper.java

```diff
@@ -0,0 +1,271 @@
+    
+                // Checking equality for refs fields.
+                for (int i = 0; i < lhStructType.getFieldTypeCount()[5]; i++) {
+                    if (!isDeepEqual(lhStruct.getRefField(i), rhStruct.getRefField(i))) {
+                        return false;
+                    }
+                }
+                break;
+            case TypeTags.MAP_TAG:
+                BMap lhMap = (BMap) lhVal;
+                BMap rhMap = (BMap) rhVal;
+                // Check if size is same.
+                if (lhMap.size() != rhMap.size()) {
+                    return false;
+                }
+    
+                // Check if key set is equal.
+                if (!lhMap.keySet().containsAll(rhMap.keySet())) {
+                    return false;
+                }
+    
+                List<String> keys = Arrays.stream(lhMap.keySet().toArray())
+                        .map(String.class::cast)
+                        .collect(Collectors.toList());
+                for (int i = 0; i < lhMap.size(); i++) {
+                    if (!isDeepEqual(lhMap.get(keys.get(i)), rhMap.get(keys.get(i)))) {
+                        return false;
+                    }
+                }
+                break;
+            case TypeTags.ARRAY_TAG:
+            case TypeTags.ANY_TAG:
+                // TODO: This block should ideally be in the ARRAY_TAG case only. Not ANY_TAG. #4505.
+                if (lhVal instanceof BNewArray) {
+                    BNewArray lhArray = (BNewArray) lhVal;
+                    BNewArray rhArray = (BNewArray) rhVal;
+                    if (lhArray.size() != rhArray.size()) {
+                        return false;
+                    }
+    
+                    for (int i = 0; i < lhArray.size(); i++) {
+                        if (!isDeepEqual(lhArray.getBValue(i), rhArray.getBValue(i))) {
+                            return false;
+                        }
+                    }
+                }
+                break;
+            case TypeTags.JSON_TAG:
+                JsonNode lhJSON = ((BJSON) lhVal).value();
+                JsonNode rhJSON = ((BJSON) rhVal).value();
+    
+                return isDeepEqual(lhJSON, rhJSON);
+            case TypeTags.XML_TAG:
+                // https://www.pixelstech.net/article/1453272563-Canonicalize-XML-in-Java
+                break;
+            case TypeTags.CONNECTOR_TAG:
+                BConnector lhConnector = (BConnector) lhVal;
+                BConnector rhConnector = (BConnector) rhVal;
+                
+                if (!lhConnector.getConnectorType().getPackagePath().equals(
+                        rhConnector.getConnectorType().getPackagePath())) {
+                    return false;
+                }
+    
+                if (!lhConnector.getConnectorType().getName().equals(rhConnector.getConnectorType().getName())) {
+                    return false;
+                }
+                break;
+            case TypeTags.ENUM_TAG:
```

Comment:
Do we really need the deep equality of enums?

Reply:
Removed support for enums. See https://github.com/ballerina-lang/ballerina/pull/4532/commits/de2bf293cdf607e12055f2e7e3bb4ab78276774d